python/ray/workflow/workflow_storage.py

```diff
@@ -86,6 +87,124 @@ def is_recoverable(self) -> bool:
         )
 
 
+class WorkflowIndexingStorage:
+    """Access and maintenance the indexing of workflow status.
+
+    It runs a protocol that guarantees we can recover from any interrupted
+    status updating. This protocol is **not thread-safe** for a workflow, currently
+    it is executed by workflow management actor with a single thread.
+
+    Here is how the protocol works:
+
+    Loading a status of a workflow
```

Comment:
maybe we don't need fixing when we are doing loading/reading. With this, we can relax the thread-safety requirement I think.

Basically when listing/loading, we just do read as we need. When writing, we'll check the things there and do the fixing.

Will this be better?

Reply:
don't do fixing when reading/loading seems invalidate the purpose of this feature request - indexing workflow status.  

And there is another code that updates the status in parallel:

```python
def cancel(workflow_id: str) -> None:
    try:
        workflow_manager = get_management_actor()
        ray.get(workflow_manager.cancel_workflow.remote(workflow_id))
    except ValueError:
        wf_store = workflow_storage.get_workflow_storage(workflow_id)
        wf_store.update_workflow_status(WorkflowStatus.CANCELED)
```

so even updating on writing is not safe. I can still make it work but it requires a much more complicated protocol. to me it seems we should always call the workflow management actor for status loading/updating.